// sql-ddl-to-json-schema/lib/mysql/constants.js
"use strict";
// https://learn.microsoft.com/en-us/sql/t-sql/data-types/int-bigint-smallint-and-tinyint-transact-sql?view=sql-server-ver16
// https://dev.mysql.com/doc/refman/8.4/en/integer-types.html
Object.defineProperty(exports, "__esModule", { value: true });
exports.BIGINT_UNSIGNED_MAX_1 = exports.BIGINT_UNSIGNED_MIN_1 = exports.BIGINT_SIGNED_MAX_1 = exports.BIGINT_SIGNED_MIN_1 = exports.MIN_SAFE_INTEGER = exports.MAX_SAFE_INTEGER = exports.BIGINT_UNSIGNED_MAX = exports.BIGINT_UNSIGNED_MIN = exports.BIGINT_SIGNED_MAX = exports.BIGINT_SIGNED_MIN = exports.INT_UNSIGNED_MAX = exports.INT_UNSIGNED_MIN = exports.INT_SIGNED_MAX = exports.INT_SIGNED_MIN = exports.MEDIUMINT_UNSIGNED_MAX = exports.MEDIUMINT_UNSIGNED_MIN = exports.MEDIUMINT_SIGNED_MAX = exports.MEDIUMINT_SIGNED_MIN = exports.SMALLINT_UNSIGNED_MAX = exports.SMALLINT_UNSIGNED_MIN = exports.SMALLINT_SIGNED_MAX = exports.SMALLINT_SIGNED_MIN = exports.TINYINT_UNSIGNED_MAX = exports.TINYINT_UNSIGNED_MIN = exports.TINYINT_SIGNED_MAX = exports.TINYINT_SIGNED_MIN = void 0;
exports.TINYINT_SIGNED_MIN = -128;
exports.TINYINT_SIGNED_MAX = 127;
exports.TINYINT_UNSIGNED_MIN = 0;
exports.TINYINT_UNSIGNED_MAX = 255;
exports.SMALLINT_SIGNED_MIN = -32768;
exports.SMALLINT_SIGNED_MAX = 32767;
exports.SMALLINT_UNSIGNED_MIN = 0;
exports.SMALLINT_UNSIGNED_MAX = 65535;
exports.MEDIUMINT_SIGNED_MIN = -8388608;
exports.MEDIUMINT_SIGNED_MAX = 8388607;
exports.MEDIUMINT_UNSIGNED_MIN = 0;
exports.MEDIUMINT_UNSIGNED_MAX = 16777215;
exports.INT_SIGNED_MIN = -2147483648;
exports.INT_SIGNED_MAX = 2147483647;
exports.INT_UNSIGNED_MIN = 0;
exports.INT_UNSIGNED_MAX = 4294967295;
// -BigInt(1) * BigInt(2) ** BigInt(63)
exports.BIGINT_SIGNED_MIN = -9223372036854775808n;
// BigInt(2) ** BigInt(63) - BigInt(1)
exports.BIGINT_SIGNED_MAX = 9223372036854775807n;
//
exports.BIGINT_UNSIGNED_MIN = 0n;
// BigInt(2) ** BigInt(64) - BigInt(1)
exports.BIGINT_UNSIGNED_MAX = 18446744073709551615n;
// https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/BigInt
/**  9007199254740991: 2 ^ 53 - 1 */
exports.MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER;
/** -9007199254740991: -1 * (2 ^ 53 - 1) */
exports.MIN_SAFE_INTEGER = Number.MIN_SAFE_INTEGER;
// a proposal for javascript to handle big int numbers
exports.BIGINT_SIGNED_MIN_1 = exports.MIN_SAFE_INTEGER;
exports.BIGINT_SIGNED_MAX_1 = exports.MAX_SAFE_INTEGER;
exports.BIGINT_UNSIGNED_MIN_1 = 0;
exports.BIGINT_UNSIGNED_MAX_1 = exports.MAX_SAFE_INTEGER;
